src/UCTNode.cpp

```diff
@@ -92,8 +93,20 @@ bool UCTNode::create_children(Network & network,
             legal_sum += raw_netlist.policy[i];
         }
     }
-    nodelist.emplace_back(raw_netlist.policy_pass, FastBoard::PASS);
-    legal_sum += raw_netlist.policy_pass;
+
+    // Don't consider pass moves if they would be immediately losing and we "shouldn't"
+    // accept that; see issue #2273. --dumbpass disables this heuristic.
+    float final_score_for_player = state.final_score();
```

Comment:
Calculating the score could be done only after other checks have passed. Even though it's cheap compared to NN evaluation it's much more expensive than the other checks. Having to count the score at every node can affect performance with fast GPU and small net.

Reply:
Ah, yes, you're right. The score check _was_ done at the end in an earlier version of this PR, but then I pulled it out into a separate variable for clarity without thinking carefully enough about the implications.

So I think the right way to do this is for FastState to grow _either_ a new method like `final_score` except that it returns the score from the current player's perspective _or_ a new argument to `final_score` that lets the caller decide whether they want that. I think the former is slightly better and will do that, but I am open to persuasion.

Reply:
I changed my mind about the right way to do it. `final_score` can now be called with no args (meaning the same as it always did) or with one arg (meaning you tell it which player's perspective you want). I think this is clearer than e.g. making up some name meaning "final score from the perspective of the current player".